File: hash_table/c/hash_table_open_addressing.c
// Hash table using the open addressing approach (linear probing).
#include <stdlib.h>
#include <string.h>

#include "hash_table_open_addressing.h"

hash_table_t *create_hash_table(void)
{
  hash_table_t *ht = (hash_table_t *)malloc(sizeof(hash_table_t));
  if (ht == NULL)
    return NULL;
  ht->n = 0;
  ht->m = 8;
  ht->array = (hash_table_entry_t *)malloc(sizeof(hash_table_entry_t) * ht->m);
  if (ht->array == NULL)
    return NULL;

  for (int i = 0; i < ht->m; i++) {
    ht->array[i].key = NULL; ht->array[i].value = 0;
  }
  return ht;
}

static int hash(const char *key, const int m)
{
  int hash = 0;

  for (int i = 0; i < strlen(key); i++) {
    hash = hash * 29 + key[i];
  }
  return abs(hash % m);
}

void insert(hash_table_t *ht, const char *key, int value)
{
  if ((ht->n / (float)ht->m) > 0.6) {
    rehash(ht);
  }

  int h = hash(key, ht->m);
  while (ht->array[h].key != NULL && ht->array[h].key != deleted) {
    if (strcmp(ht->array[h].key, key) == 0) {
      ht->array[h].value = value;
      return;
    }
    h = (h + 1) % ht->m;
  }

  ht->array[h].key = strdup(key);
  ht->array[h].value = value;
  ht->n++;
}

int get(hash_table_t *ht, const char *key)
{
  int h = hash(key, ht->m);

  if (ht->array[h].key == NULL || ht->array[h].key == deleted)
    return -1;

  while (ht->array[h].key != NULL) {
    if (strcmp(ht->array[h].key, key) == 0)
      return ht->array[h].value;

    h = (h + 1) % ht->m;
  }
  return -1;
}

void delete(hash_table_t *ht, const char *key)
{
  int h = hash(key, ht->m);

  if (ht->array[h].key == NULL || ht->array[h].key == deleted)
    return;

  while (ht->array[h].key != NULL) {
    if (strcmp(ht->array[h].key, key) == 0) {
      free(ht->array[h].key);
      ht->array[h].key = deleted;
      ht->n--;
      return;
    }
    h = (h + 1) % ht->m;
  }
}

void rehash(hash_table_t *ht)
{
  hash_table_entry_t *array = (hash_table_entry_t *)malloc(sizeof(hash_table_entry_t) * ht->m * 2);
  if (array == NULL)
    return;

  for (int i = 0; i < ht->m * 2; i++) {
    array[i].key = NULL; array[i].value = 0;
  }

  int h;
  for (int i = 0; i < ht->m; i++) {
    if (ht->array[i].key != NULL && ht->array[i].key != deleted) {
      h = hash(ht->array[i].key, ht->m * 2);

      while (array[h].key != NULL) {
        h = (h + 1) % (ht->m * 2);
      }

      array[h].key = ht->array[i].key;
      array[h].value = ht->array[i].value;
    }
  }

  ht->m *= 2;
  free(ht->array);
  ht->array = array;
}

File: hash_table/python/hash_table_open_addressing.py
#!/usr/bin/env python2
# Hash table using the open addressing approach (linear probing).

from __future__ import division


class HashTableEntry(object):
  def __init__(self, key=None, value=None):
    self.key = key
    self.value = value


class HashTable(object):
  def __init__(self, size=8):
    self.array = [HashTableEntry() for _ in xrange(size)]
    self.m = size
    self.n = 0

  def __setitem__(self, key, value):
    return self.insert(key, value)

  def __getitem__(self, key):
    return self.get(key)

  def __delitem__(self, key):
    return self.delete(key)

  def __str__(self):
    return '\n'.join(['%s: %s' % (entry.key, entry.value) for entry in self.array])

  def _hash(self, key):
    return hash(key) % self.m

  def _resize(self):
    self.m *= 2  # table doubling
    array = [HashTableEntry() for _ in xrange(self.m)]

    for entry in self.array:
      if entry.key is not None and entry.key != '<dummy>':
        h = self._hash(entry.key)
        while array[h].key is not None:
          h = (h + 1) % self.m

        array[h] = entry

    self.array = array

  def insert(self, key, value):
    if self.n / self.m > 0.6:  # load factor
      self._resize()

    h = self._hash(key)

    while self.array[h].key is not None and self.array[h].key != '<dummy>':
      if self.array[h].key == key:
        self.array[h].value = value
        return
      h = (h + 1) % self.m

    self.array[h] = HashTableEntry(key, value)
    self.n += 1

  def get(self, key):
    h = self._hash(key)

    if self.array[h].key is None:
      raise KeyError

    while self.array[h].key is not None:
      if self.array[h].key == key:
        return self.array[h].value
      h = (h + 1) % self.m
    else:
      raise KeyError

  def delete(self, key):
    h = self._hash(key)

    if self.array[h].key is None:
      raise KeyError

    while self.array[h].key is not None:
      if self.array[h].key == key:
        self.array[h].key = '<dummy>'
        return
      h = (h + 1) % self.m
    else:
     raise KeyError


if __name__ == '__main__':
  hash_table = HashTable(16)
  hash_table[0] = 'a'
  hash_table[0 + 16] = 'b'
  # hash_table[0 + 32] = 'c'

  print hash_table
  print hash_table[0 + 32]
